src/Features/Core/Portable/SolutionCrawler/SolutionCrawlerProgressReporter.cs

```diff
@@ -19,109 +16,69 @@ internal partial class SolutionCrawlerRegistrationService : ISolutionCrawlerRegi
 
-            private int _count;
+            public bool InProgress => _progressStartCount > 0;
 
-            public SolutionCrawlerProgressReporter(IAsynchronousOperationListener listener)
-            {
-                _listener = listener;
-                _eventQueue = new SimpleTaskQueue(TaskScheduler.Default);
-                _eventMap = new EventMap();
+            public void Start() => ChangeProgressStatus(ref _progressStartCount, ProgressStatus.Started);
+            public void Stop() => ChangeProgressStatus(ref _progressStartCount, ProgressStatus.Stoped);
+
+            private void Evaluate() => ChangeProgressStatus(ref _progressEvaluateCount, ProgressStatus.Evaluating);
+            private void Pause() => ChangeProgressStatus(ref _progressEvaluateCount, ProgressStatus.Paused);
 
-                _count = 0;
-            }
 
-            public bool InProgress
+            public void UpdatePendingItemCount(int pendingItemCount)
             {
-                get
+                if (_progressStartCount > 0)
                 {
-                    return _count > 0;
+                    var progressData = new ProgressData(ProgressStatus.PendingItemUpdated, pendingItemCount);
+                    OnProgressChanged(progressData);
                 }
             }
 
-            public event EventHandler<ProgressData> ProgressChanged
+            public IDisposable Evaluating()
             {
-                add
-                {
-                    _eventMap.AddEventHandler(nameof(ProgressChanged), value);
-                }
-
-                remove
-                {
-                    _eventMap.RemoveEventHandler(nameof(ProgressChanged), value);
-                }
+                return new ProgressStatusRAII(this);
             }
 
-            public Task Start()
+            private void ChangeProgressStatus(ref int referenceCount, ProgressStatus status)
             {
-                if (Interlocked.Increment(ref _count) == 1)
+                var start = (status == ProgressStatus.Started || status == ProgressStatus.Evaluating);
+                if (start ? (Interlocked.Increment(ref referenceCount) == 1) : (Interlocked.Decrement(ref referenceCount) == 0))
                 {
-                    _lastReportedFilePath = null;
-
-                    var asyncToken = _listener.BeginAsyncOperation("ProgressReportStart");
-                    var progressData = new ProgressData(ProgressStatus.Started, filePathOpt: null);
-                    return RaiseEvent(nameof(ProgressChanged), progressData).CompletesAsyncOperation(asyncToken);
+                    var progressData = new ProgressData(status, pendingItemCount: null);
+                    OnProgressChanged(progressData);
                 }
-
-                return Task.CompletedTask;
             }
 
-            public Task Stop()
+            private void OnProgressChanged(ProgressData progressData)
```

Comment:
considering inlining this.  i don't think it adds much value.

Reply:
also I dont believe there is any harm.